hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/preloader/NearCachePreloader.java

```diff
@@ -0,0 +1,303 @@
+     * @param <NCRM>  the {@link NearCacheRecordMap} type
+     */
+    public <R extends NearCacheRecord, NCRM extends NearCacheRecordMap<KS, R>> void storeKeys(NCRM... records) {
+        long started = System.nanoTime();
+        FileOutputStream fos = null;
+        try {
+            tmpBytes = new byte[INT_SIZE_IN_BYTES];
+            lastWrittenBytes = 0;
+            lastKeyCount = 0;
+
+            fos = new FileOutputStream(tmpStoreFile, false);
+
+            // writer header
+            writeInt(fos, MAGIC_BYTES);
+            writeInt(fos, FileFormat.INTERLEAVED_LENGTH_FIELD.ordinal());
+
+            // writer keys
+            for (NCRM record : records) {
+                writeKeySet(fos, fos.getChannel(), record.keySet());
+            }
+
+            // cleanup if no keys have been written
+            if (lastKeyCount == 0) {
+                deleteQuietly(storeFile);
+                persistenceStats(started);
+                return;
+            }
+
+            fos.flush();
+            rename(tmpStoreFile, storeFile);
+
+            persistenceStats(started);
+        } catch (Exception e) {
+            logger.warning(format("Could not store keys of Near Cache %s (%s): [%s] %s", nearCacheName,
+                    storeFile.getAbsolutePath(), e.getClass().getSimpleName(), e.getMessage()));
+        } finally {
+            tmpBytes = null;
+            deleteQuietly(tmpStoreFile);
+            closeResource(fos);
+        }
+    }
+
+    private void persistenceStats(long started) {
+        long elapsedMillis = getElapsedMillis(started);
+        nearCacheStats.addPersistence(elapsedMillis, lastWrittenBytes, lastKeyCount);
+
+        logger.info(format("Stored %d keys of Near Cache %s in %d ms (%d kB)", lastKeyCount, nearCacheName, elapsedMillis,
+                MemoryUnit.BYTES.toKiloBytes(lastWrittenBytes)));
+    }
+
+    private int loadKeySet(BufferingInputStream bis, DataStructureAdapter<Data, ?> adapter) throws IOException {
+        int loadedKeys = 0;
+
+        Builder<Data> builder = InflatableSet.newBuilder(LOAD_BATCH_SIZE);
+        while (readFullyOrNothing(bis, tmpBytes)) {
+            int dataSize = readIntB(tmpBytes, 0);
+            byte[] payload = new byte[dataSize];
+            if (!readFullyOrNothing(bis, payload)) {
+                break;
+            }
+            builder.add(new HeapData(payload));
+            if (builder.size() == LOAD_BATCH_SIZE) {
+                adapter.getAll(builder.build());
+                builder = InflatableSet.newBuilder(LOAD_BATCH_SIZE);
+            }
+            loadedKeys++;
+        }
+        if (builder.size() > 0) {
+            adapter.getAll(builder.build());
+        }
+        return loadedKeys;
+    }
+
+    private void writeKeySet(FileOutputStream fos, FileChannel outChannel, Set<KS> keySet) throws IOException {
+        for (KS key : keySet) {
+            Data dataKey = serializationService.toData(key);
+            if (dataKey != null) {
```

Comment:
I have no test case for this (since it's not so easy to reproduce), so we would have an untested code branch in the code. In those cases I sometimes write down the "happy path" to have the lines covered.